src/main/scala/akka/stream/contrib/Retry.scala

```diff
@@ -0,0 +1,245 @@
+ */
+object Retry {
+
+  /**
+   * Factory for a retry flow. Given a flow which recieves a tuple of "input" & "state",
+   * and emits a tuples of a Try of "output" & "state", and given a function that given a state
+   * can produce a new tuple of "input-state" to pass again through the flow, this function will
+   * yield a new flow, which wraps the given flow, and retries failed (i.e. elements passed through
+   * the flow and come out with `Failure[O]` as the first part of the tuple) elements with the new
+   * "input-state" tuple recieved from the `retryWith` function, and outputs a `Try[O]`.
+   * To not continue retrying (and getting a `Failure[O]` as output), the `retryWith` function
+   * should return a `None`
+   *
+   * IMPORTANT CAVEAT:
+   * The given flow must not change the number of elements passing through it (i.e. it should output
+   * exactly one element for every recieved element). Ignoring this, will have an unpredicted result.
+   *
+   * @param flow the flow to retry
+   * @param retryWith if output was failure, we can optionaly recover from it,
+   *        and retry with a new pair of input & new state we get from this function.
+   * @tparam I input elements type
+   * @tparam O output elements type
+   * @tparam S state to create a new (input,state): (I,S) to retry with
+   * @tparam M materialized value type
+   * @return `Graph[FlowShape[(I,S),Try[O]],M]` instance
+   */
+  def apply[I, O, S, M](flow: Graph[FlowShape[(I, S), (Try[O], S)], M])(retryWith: S => Option[(I, S)]): Graph[FlowShape[(I, S), Try[O]], M] = {
+    GraphDSL.create(flow) { implicit b => origFlow =>
+      import GraphDSL.Implicits._
+
+      val elemCount = b.add(new CountElementsAndInterceptCompletion[(I, S)])
+      val mergePref = b.add(MergePreferred[(I, S)](1, true))
+      val partition = b.add(PartitionWith[(Try[O], S), Try[O], (I, S)] {
+        case (success: Success[O], _) => Left(success)
+        case (failedTry, s)           => retryWith(s).fold[Either[Try[O], (I, S)]](Left(failedTry))(Right.apply)
+      })
+      val coElemCount = b.add(new InterceptElemntsLeavingCycle[Try[O]])
+      // format: OFF
+      elemCount.out ~> mergePref ~> origFlow ~> partition.in
+                                                partition.out0 ~> coElemCount.in
+                       mergePref.preferred  <~  partition.out1
+      elemCount.in1                         <~                    coElemCount.out1
+      // format: ON
+      FlowShape(elemCount.in0, coElemCount.out0)
+    }
+  }
+
+  /**
+   * Factory for multiple retries flow. similar to the simple retry, but this will allow to
+   * break down a "heavy" element which failed into multiple "thin" elements, that may succeed
+   * indevidually. Since it's very easy to inflate elements in retry cycle, there's also a
+   * limit parameter, that will limit the amount of generated elements by the `retryWith` function,
+   * and will fail the stage if that limit is exceeded.
```

Comment:
Maybe the API can be expanded to receive an overflow strategy, like the buffer stage.
I have thought of that, and decided it's an overkill for now.  But since it's the contrib, I assume that if down the road we break API and add another overflow strategy parameter, it'll be OK.

anyway, would you want me to add that now before accepting this PR?

P.S. as a default behaviour, IMO, failing the stage is better. as a user of this library, I would want to know in the development phase I did something wrong, instead of discovering in production I'm losing data silently.